skorch/hf.py

```diff
@@ -0,0 +1,800 @@
+
+    post_processor : tokenizers.processors.PostProcessor
+      Optional post-processor, mostly used to add special tokens for BERT etc.
+
+    max_length : int (default=256)
+      Maximum number of tokens used per sequence.
+
+    return_tensors : one of None, str, 'pt', 'np', 'tf' (default='pt')
+      What type of result values to return. By default, return a padded and
+      truncated (to ``max_length``) PyTorch Tensor. Similarly, 'np' results in a
+      padded and truncated numpy array. Tensorflow tensors are not supported
+      officially supported but should also work. If None or str, return a list
+      of lists instead. These lists are not padded or truncated, thus each row
+      may have different numbers of elements.
+
+    return_attention_mask : bool (default=True)
+      Whether to return the attention mask.
+
+    return_token_type_ids : bool (default=False)
+      Whether to return the token type ids.
+
+    return_length : bool (default=False)
+      Whether to return the length of the encoded inputs.
+
+    pad_token : str (default='[PAD]')
+      A special token used to make arrays of tokens the same size for batching
+      purpose. Will then be ignored by attention mechanisms.
+
+    verbose : int (default=0)
+      Whether the tokenizer should print more information and warnings.
+
+    Attributes
+    ----------
+    vocabulary_ : dict
+      A mapping of terms to feature indices.
+
+    fast_tokenizer_ : transformers.PreTrainedTokenizerFast
+      If you want to extract the Huggingface tokenizer to use it without skorch,
+      use this attribute.
+
+    .. _tokenizers: https://huggingface.co/docs/tokenizers/python/latest/index.html
+
+    """
+    import transformers as _transformers
+    import tokenizers as _tokenizers
+
+    prefixes_ = [
+        'model', 'normalizer', 'post_processor', 'pre_tokenizer', 'tokenizer', 'trainer'
+    ]
+
+    def __init__(
+            self,
+            tokenizer,
+            model=None,
+            trainer='auto',
+            normalizer=None,
+            pre_tokenizer=None,
+            post_processor=None,
+            max_length=256,
+            return_tensors='pt',
+            return_attention_mask=True,
+            return_token_type_ids=False,
+            return_length=False,
+            pad_token='[PAD]',
+            verbose=0,
+            **kwargs,
+    ):
+        self.tokenizer = tokenizer
+        self.model = model
+        self.trainer = trainer
+        self.normalizer = normalizer
+        self.pre_tokenizer = pre_tokenizer
+        self.post_processor = post_processor
+        self.max_length = max_length
+        self.return_tensors = return_tensors
+        self.return_attention_mask = return_attention_mask
+        self.return_token_type_ids = return_token_type_ids
+        self.return_length = return_length
+        self.pad_token = pad_token
+        self.verbose = verbose
+
+        self._kwargs = kwargs
+        vars(self).update(kwargs)
```

Comment:
Yes, it's not something you would normally see in sklearn. However, I don't know if there is a better way to make it possible to have arbitrary parameters, e.g. `pre_tokenizer__delimiter`? I guess we could hide it inside `_kwargs` but having it only there can create its own problems, e.g. if a user tries to access `my_tokenizer.pre_tokenizer__delimiter` (instead of usng `get_params`) or if they set `my_tokenizer.pre_tokenizer__delimiter = 'foo'` directly (instead of using `set_params`). Yes, this is all fixable with some Python magic but I think having the attributes exist on the object directly is the more transparent and less error prone way.

Reply:
Are these parameters of sub-estimators? `pre_tokenizer__delimiter` seems like one to me. In that case, you could maybe use `_BaseComposition` maybe? https://github.com/scikit-learn/scikit-learn/blob/6e7922e48a04f471b11e874fac67657941970794/sklearn/utils/metaestimators.py#L23 (and the usage in `Pipeline`)

Reply:
The tokenizer is a sub-component, though not a `BaseEstimator`. In skorch, we stick close to `_BaseComposition._set_params` but have a few modifications to make certain things work, so we cannot just use `_BaseComposition`.